awx/main/models/notifications.py

```diff
@@ -201,56 +247,145 @@ def get_absolute_url(self, request=None):
 
 
 class JobNotificationMixin(object):
+    STATUS_TO_TEMPLATE_TYPE = {'succeeded': 'success',
+                               'running': 'started',
+                               'failed': 'error'}
+    # Tree of fields that can be safely referenced in a notification message
+    JOB_FIELDS_WHITELIST = ['id', 'type', 'url', 'created', 'modified', 'name', 'description', 'job_type', 'inventory',
+                            'project', 'playbook', 'forks', 'limit', 'verbosity', 'job_tags', 'force_handlers', 'skip_tags',
+                            'start_at_task', 'timeout', 'use_fact_cache', 'unified_job_template', 'launch_type', 'status',
```

Comment:
General comments:
- Is there a reason to whitelist both `unified_job_template` and `job_template`?
- attempting to template out `event_processing_finished` would potentially be a race condition if we're not careful
- an argument could be made that we don't need the unitary PKs for `inventory`, `project`, etc. when we have the summary fields
- are job_host_summaries a possibility?

Reply:
* Workflows???

Reply:
Yea, I'd probably remove `event_processing_finished`, and the ones that are primary keys, since we've got summary field data.

Reply:
- [x] Removed `event_processing_finished`
- [x] Removed pk fields already included in summary fields
- [ ] workflows

Reply:
> are job_host_summaries a possibility?

@wenottingham - you mean, could you write something like this in a notification message?
```
{{ job.name }} is based on JT '{{ job.summary_fields.job_template.name }}', inventory '{{ job.summary_fields.inventory.name }}', and project '{{ job.summary_fields.project.name}}'
```

If that's what you're asking, you definitely can use nested fields like that:

![image](https://user-images.githubusercontent.com/4440360/62099399-be315300-b242-11e9-8058-3dd5b2937bd4.png)

---

I checked up on the question marks shown in the above screenshot. Confirmed that the question marks are explicitly / intentionally inserted by towerkit whenever `fauxfactory` returns a unicode character that is outside of the basic multilingual plane (BMP):
https://github.com/ansible/towerkit/blob/236ed7f2b74863c6e0f22fc07611dc273fce5486/towerkit/utils.py#L236

![image](https://user-images.githubusercontent.com/4440360/62099643-a3aba980-b243-11e9-9c64-79da45ea88d3.png)

In other words, we're not seeing missing / malformed unicode characters as a result of the templating we're doing. Did some explicit testing just to triple-check:

![image](https://user-images.githubusercontent.com/4440360/62099768-fbe2ab80-b243-11e9-8ebf-c44a3325f22e.png)

![image](https://user-images.githubusercontent.com/4440360/62099790-0b61f480-b244-11e9-94c5-dd9fac8b80f6.png)

Reply:
I mean the related job_host_summaries field, i.e., `https://tower-35/api/v2/jobs/80/job_host_summaries/`

The use case for this is that it's the only way to get a view of what hosts specifically failed.